domain/port/state/watcher_test.go

```diff
@@ -0,0 +1,117 @@
+	"context"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/collections/set"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/domain"
+	machinestate "github.com/juju/juju/domain/machine/state"
+	"github.com/juju/juju/internal/logger"
+)
+
+type watcherSuite struct {
+	baseSuite
+
+	unitUUIDs [3]string
+
+	appUUIDs [2]string
+}
+
+var _ = gc.Suite(&watcherSuite{})
+
+var (
+	ssh   = network.PortRange{Protocol: "tcp", FromPort: 22, ToPort: 22}
+	http  = network.PortRange{Protocol: "tcp", FromPort: 80, ToPort: 80}
+	https = network.PortRange{Protocol: "tcp", FromPort: 443, ToPort: 443}
+)
+
+func (s *watcherSuite) SetUpTest(c *gc.C) {
+	s.ModelSuite.SetUpTest(c)
+
+	machineSt := machinestate.NewState(s.TxnRunnerFactory(), logger.GetLogger("juju.test.machine"))
+
+	err := machineSt.CreateMachine(context.Background(), "0", netNodeUUIDs[0], machineUUIDs[0])
+	c.Assert(err, jc.ErrorIsNil)
+
+	s.unitUUIDs[0], s.appUUIDs[0] = s.createUnit(c, netNodeUUIDs[0], appNames[0])
+	s.unitUUIDs[1], s.appUUIDs[1] = s.createUnit(c, netNodeUUIDs[0], appNames[1])
+
+	err = machineSt.CreateMachine(context.Background(), "1", netNodeUUIDs[1], machineUUIDs[1])
+	c.Assert(err, jc.ErrorIsNil)
+
+	s.unitUUIDs[2], _ = s.createUnit(c, netNodeUUIDs[1], appNames[1])
+}
+
+func (s *watcherSuite) initialiseOpenPorts(c *gc.C, st *State) ([]string, map[string]string) {
+	err := st.RunAtomic(context.Background(), func(ctx domain.AtomicContext) error {
+		err := st.UpdateUnitPorts(ctx, s.unitUUIDs[0], network.GroupedPortRanges{
+			"ep0": {ssh},
+		}, network.GroupedPortRanges{})
+		if err != nil {
+			return err
+		}
+		err = st.UpdateUnitPorts(ctx, s.unitUUIDs[1], network.GroupedPortRanges{
+			"ep1": {http},
+		}, network.GroupedPortRanges{})
+		if err != nil {
+			return err
+		}
+		return st.UpdateUnitPorts(ctx, s.unitUUIDs[2], network.GroupedPortRanges{
+			"ep2": {https},
+		}, network.GroupedPortRanges{})
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	query, err := sqlair.Prepare(`
+SELECT &endpoint.*
+FROM unit_endpoint
+`, endpoint{})
+	c.Assert(err, jc.ErrorIsNil)
+
+	var endpoints []endpoint
+	err = s.TxnRunner().Txn(context.Background(), func(ctx context.Context, tx *sqlair.TX) error {
+		return tx.Query(ctx, query).GetAll(&endpoints)
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	endpointToUUIDMap := make(map[string]string)
+	endpointUUIDs := make([]string, len(endpoints))
+	for i, ep := range endpoints {
+		endpointToUUIDMap[ep.Endpoint] = ep.UUID
+		endpointUUIDs[i] = ep.UUID
+	}
+
+	return endpointUUIDs, endpointToUUIDMap
+}
+
+func (s *watcherSuite) TestGetMachinesForPortRanges(c *gc.C) {
+	st := NewState(s.TxnRunnerFactory())
+	ctx := context.Background()
+	endpointUUIDs, _ := s.initialiseOpenPorts(c, st)
+
+	machineUUIDs, err := st.GetMachinesForEndpoints(ctx, endpointUUIDs)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(machineUUIDs, jc.SameContents, []string{machineUUIDs[0], machineUUIDs[1]})
```

Comment:
**question(what???):** Maybe I am not seeing something REALLY obvious, but I don't see the difference between this check, and `c.Check(machineUUIDs, gc.HasLen, 2)`, logically-wise (they would fails the same, but len check is just more readable)

Either I miss something in the purpose of the test, or my logic is wrong on Thursday (one day earlier than usual :D) , or there is something wrong here. Please double check.